### DISCUSSION TITLE
[10 Giugno 2022] Assembly

### DISCUSSION THREAD

**Discussion by Elia-Belli** [sapienzastudentsnetwork/architettura-degli-elaboratori#47] [Assembly (M-Z)]
![Esame](https://github.com/sapienzastudentsnetwork/archelab2223/assets/107944476/e69f0e16-cdfb-4c69-b7f1-e942b3ec3382)
---
### Soluzione del Prof
Le soluzioni fornite dal professore potrebbero essere incomplete o poco chiare, quindi sono benvenute soluzioni che le completano, oppure soluzioni alternative e/o con spiegazioni più approfondite: [Soluzione.txt](https://github.com/sapienzastudentsnetwork/archelab2223/files/11666693/Soluzione.txt)

**Comment by daromacarlo** [comment]
```
.globl main
.data
 S: .ascii "0110101011\n"
 C: .ascii "\n"
.text 
main:
	la $t1,C
	la $a0,S
	lb $t2,($t1)		
	contaOccorrenze:
	
	jal while 	#"jumpo" alla label while e salvo nel registro $ra l'istruzione successiva a questa
	
	#chiamate a sistema
	move $a0,$v0
	li $v0,1
	syscall
	move $a0,$t1
	li $v0,4
	syscall
	move $a0,$v1
	li $v0,1
	syscall
	li $v0,10
	syscall
	
	while:
	lb $t0,($a0)
	addi $a0,$a0,1
	beq $t0,$t2,fine
	subi $t0,$t0,48 
			#	dato che "0" in ascii è 110000
			#	quindi sottraggo 48
	bne $t0,$0,uno
	addi $v0,$v0,1  #contatore occorrenze 0 sale di 1
	j while		#torno all'inizio del ciclo
	
	uno:
	
	addi $v1,$v1,1  #contatore occorrenze 0 sale di 0
	j while		#contatore occorrenze 1 sale di 1
	
	fine:
	jr $ra		#torno alla riga dopo il comando jal
``` 


#programma da sistemare, manca la parte per #prendere l'input da utente.
#Tuttavia la logica dovrebbe essere giusta

**Comment by daromacarlo** [comment]
```
################ approccio ricorsivo #####################################################

.data
    buffer: .space 20
.text

main:
    li $v0, 8       # Codice per input stringa
    la $a0, buffer  # Carica indirizzo base in $a0
    li $a1, 20      # Alloca al massimo 20 caratteri
    syscall         # $a0 contiene l'indirizzo base della stringa
	
    li $v0,0
    jal contaOccRec
	
	move $a0,$v0
	li $v0,1
	syscall
	
	li $a0,'\n'
	li $v0,11
	syscall
	
	move $a0,$v1
	li $v0,1
	syscall
	
	li $v0,10
	syscall
	
##########################################
## INSERIRE IL CODICE QUI

	contaOccRec: 
	
		li $t0, '\n'	#carico il carattere che fa terminare la stringa 
		lb $t1,($a0)	#carico in $t1 il carattere
		beq $t1,$t0,casobase #se sono arrivato all'ultimo carattere salto al caso base
		
		subi $sp,$sp,12  #carico lo stack
		sw $t1,8($sp)
		sw $ra,4($sp)
		sw $a0,($sp)
		addi $a0,$a0,1 #il puntatore sale di un byte per scorrere il vettore
		
		jal contaOccRec
		
		lw $t1,8($sp)
		lw $ra,4($sp)
		lw $a0,($sp)
		addi $sp,$sp,12  #svuoto lo stack
		
		subi $t1,$t1,48 #controllo se il nuemro è 1 o 0
		beq $t1,$0,zero #se è zero salto a zero e faccio salire il contatore di 0
	        addi $v1,$v1,1	#altrimenti faccio salire il contatore di 1
	        jr $ra
	        
		zero:
		
		addi $v0,$v0,1
		jr $ra
		
	casobase:
	
		jr $ra
```

**Comment by FeddyLix17** [comment]
@CarloDaRomadev ho provato a testare entrambi i tuoi codici con i test case della cartella di questo esercizio e questo è ciò che ho riscontrato:

**Soluzione Ricorsiva**:
- il codice viene eseguito senza interruzioni ma a schermo non viene stampato nulla (non so dire se la soluzione sia corretta in quanto non ho grande conoscenza della ricorsione in *MIPS*, a occhio l'approccio sembra giusto)

**Soluzione Iterativa**:
- il codice non viene proprio compilato e analizzandolo mi sono sorte alcune domande
1. come mai non hai messo la syscall per prende l'input? (mentre nella soluzione ricorsiva è stata implementata correttamente)

```asm
# gestire una singola sequenza S           # gestire una qualsiasi sequenza data in input
.data                                     .data
 S: .ascii "0110101011\n"                     buffer: .space 20
 C: .ascii "\n"                           .text
                                                main:
                                                   li $v0, 8       # Codice per input stringa
                                                   la $a0, buffer  # Carica indirizzo base in $a0
                                                   li $a1, 20      # Alloca al massimo 20 caratteri
                                                   syscall         # $a0 contiene l'indirizzo base della stringa
```
2. in queste 3 righe
```asm
la $t1,C
la $a0,S
lb $t2,($t1)
```
non conveniva direttamente fare un `lb $t2, C` o anche un `li $t2, '\n'`? (non so se fossero queste 3 righe o altro a non far compilare il programma), anche qui comunque l'approccio avuto di seguito sembra corretto


**allego comunque di seguito la mia soluzione *iterativa* che passa gli attuali test-case, visionabile [qui](https://github.com/FedVlogger17/Uni-Notes/blob/main/Primo%20Anno/Secondo%20Semestre/Architettura%20degli%20Elaboratori/Esame%20Mips%2010%20Giugno%202022/program01.asm)**

**Comment by daromacarlo** [comment]
@FeddyLix17  , Ciao, grazie per aver provato i miei programmi, rispondo alle due domande. Nella soluzione iterativa non ho preso l'input da syscall perchè era uno dei primi esercizi che facevo in assoluto e non sapevo dovessi farlo, dato che nella richiesta il prof non lo dice, mi sono dimenticato di modificarlo in seguito, il codice penso che tuttavia funzioni, visionerò il tuo. Per quanto riguarda invece la soluzione ricorsiva l'ho ritestata e sembra filare tutto liscio, allego video.

https://github.com/sapienzastudentsnetwork/archelab2223/assets/135441131/1f57a603-0c21-4231-8767-4c4605c1f734

**Comment by FeddyLix17** [comment]
@CarloDaRomadev grazie per la risposta e per il video, ho riprovato la tua sua soluzione ricorsiva e passa anch'essa tutti i test-case (probabilmente l'avevo ricopiata male inizialmente🙏)

**Comment by daromacarlo** [comment]
@FeddyLix17 no problem, prego 👍👍
